### ISSUE TITLE
机器学习之过拟合（overfitting）

### ISSUE THREAD

**Issue by towerjoo** [towerjoo/myML#3]
## 什么是过拟合(overfitting)

ML一个很重要的假设是输入数据集和测试数据集是independent, identical distributed(i.i.d)，即独立同分布，这个很好理解，因为倘若二者不是i.i.d那么就不存在一个相同的pattern通过机器学习习得，机器学习的目标：对于新数据的预测便成为不可能。

有了这个前提，使得机器学习成为可能（也就是找到那个相同的pattern，或者近似），但是这个pattern是未知的，对于未知的东西我们无法精确的表达，或者说我们只有知道所有这个pattern的数据集才能精确地确定这个pattern的表达，于是我们使用的输入数据显然是不完备的，基于不完备信息得出的pattern也自然是不准确的，这种不准确表达出来就是两种，一种是underfitting（欠拟合），另一种是过拟合（overfitting）。

underfitting通常是针对训练集，例如我们的预测pattern过于平滑（e.g直线来拟合多个输入点），也就是模型过于简单，这个通常可以通过引入多项式，或者神经网络等复杂模型来解决。

overfitting则通常是说训练习得的model对于测试集性能很差，而对训练集性能很好，也就是对训练集是overfitting，有很差的generalization（泛化能力），测试集的数据是我们的模型不曾见过的（参数无法基于此来训练），虽然我们有i.i.d来保证训练集和测试集在分布上不会偏差太大，但是当样本有限时偏差肯定是不可避免的，特别是当训练集很小时，训练习得的pattern可能与目标pattern差别很大。

## 为什么会出现过拟合

训练集的不完备性是导致出现拟合问题的主要原因，而数据的不完备是不可避免的（如果我们的训练集是完备的，具有所有pattern下的数据，那么整个问题就不是learning problem，而是test problem，我们其实是验证而不是学习），因此出现拟合问题也是不可避免的。

另外我们上面提到的underfitting，通常可以通过使用复杂的model来改善，但是当我们的Model解决了underfitting的问题时，也就是说在训练集的表现(performance)非常好时，并不能保证这个model在测试集表现同样好，而通常过于复杂的model会出现在训练集变现很好（甚至0 error)但是在测试集表现很差的问题。

![overfitting](https://cloud.githubusercontent.com/assets/145304/23832048/5dbffeb8-0768-11e7-9f07-f3582f912a32.png)

如上图，training error和test error(generilization error)通常会随着Model的复杂度有个先向下再向上的一个变化过程，过于复杂的model通常会导致overfitting.

![overfitting2](https://cloud.githubusercontent.com/assets/145304/23832088/37622a9c-0769-11e7-86a6-a5f93ee9f36b.png)

如上图是error与m(训练集大小）关系的曲线，我们会发现训练集越大（信息越完备），通常会有更小的test error，图中的J(theta)是所谓的loss function（参考 #1 ），loss function的大小等价于相应数据集的error. （是否有想起有人说过机器学习的核心其实不是model，而是data）

所以直观地说，过于复杂的model通常会导致overfitting，而数据集越大可以缓解overfitting.

## 如何避免过拟合

其实谈到underfitting, overfitting, generalization，肯定是逃不开bias(偏差）和variance（方差）的，周老师的书里P44有较详细的介绍。其中bias衡量的是我们model预测的值与实际值的偏差，换句话说就是拟合程度；variance指的是训练集的变动导致的学习性能的变动，即数据扰动所造成的影响（回想线性几何中方差是用来衡量随机变量和期望之间的偏离程度）。

![bias](https://cloud.githubusercontent.com/assets/145304/23832255/3fe561ae-076c-11e7-993a-9d8b5b17db98.png)

那么如何避免过拟合呢？

通常有如下几种办法：

1. 正规化（regularization)：也就是在loss function中增加一个regularization term来平衡generalization和overfitting，通常这个term为 lambda w^T w，lambda是正则变量，也是一个hyperparameter，是我们自己来设定的。有了这个term，我们可以发现，lambda越大，这个term的权重也越大，那么相应的w参数权重越小，也就是曲线越平滑，极端情况，lambda很大，那么w的各个参数趋近于0，从而使得pattern接近于一条直线（underfitting)；lambda越小，整个term的权重也越小，相应的w的权重越大，曲线则越不平滑，极端情况lambda=0则退化为没有正规化的loss function，从而overfitting. lambda这样的hyperparameter的选取通常是得不到数学的保障的，所以为什么很多人将ML的model选取和古代道士炼丹一样（更多的是经验和运气）。当然我们可以plot不同的lambda的选择形成的loss function，从而选择合适的lambda（似乎比道士高不了多少）
2. 增加训练数据集：Andrew Ng老师提到过我们需要不怕dirty的工作，而数据的准备就是这样的dirty work。我们可以手工标记新的数据，自动生成数据等等，都可以增加训练集的样本数，从而在一定程度上缓解overfitting.

## 最后

一件事情没有数学背书其实是很让人担忧的，就像著名的有人研究数学的完整性和一致性大半辈子，最终哥德尔来个哥德尔不完备定理让很多年的心血化为泡影。当然ML的我们大可不必过于担心，我们可以看到DL在各个领域的摧城拔寨，而我们的Andrew Ng老师甚至将其视为电一般的发明与影响，我们大可放开手脚大干一场，但是隐隐的数学的更多严格支持也是需要我们关注的。

## 参考文献

1. Deep Learning book
2. Learning from data(Caltech)
3. 周志华 《机器学习》
